python/hail2/matrix.py

```diff
@@ -0,0 +1,778 @@
+                    raise ExpressionException(
+                        'Type mismatch for Matrix col key: expected key type {}, found {}'.format(
+                            str(self.colkey_schema), str(col_key._type)))
+
+            if row_key is not None and col_key is not None:
+                return self.index_entries(row_key, col_key)
+            elif row_key is not None and col_key is None:
+                return self.index_rows(row_key)
+            elif row_key is None and col_key is not None:
+                return self.index_cols(col_key)
+            else:
+                return self.index_globals()
+
+    @property
+    def _jvdf(self):
+        if self._jvdf_cache is None:
+            self._jvdf_cache = Env.hail().variant.VariantDatasetFunctions(self._jvds)
+        return self._jvdf_cache
+
+    @property
+    def global_schema(self):
+        if self._global_schema is None:
+            self._global_schema = Type._from_java(self._jvds.globalSignature())
+        return self._global_schema
+
+    @property
+    def colkey_schema(self):
+        if self._colkey_schema is None:
+            self._colkey_schema = Type._from_java(self._jvds.sSignature())
+        return self._colkey_schema
+
+    @property
+    def col_schema(self):
+        if self._sa_schema is None:
+            self._sa_schema = Type._from_java(self._jvds.saSignature())
+        return self._sa_schema
+
+    @property
+    def rowkey_schema(self):
+        if self._rowkey_schema is None:
+            self._rowkey_schema = Type._from_java(self._jvds.vSignature())
+        return self._rowkey_schema
+
+    @property
+    def row_schema(self):
+        if self._va_schema is None:
+            self._va_schema = Type._from_java(self._jvds.vaSignature())
+        return self._va_schema
+
+    @property
+    def entry_schema(self):
+        if self._genotype_schema is None:
+            self._genotype_schema = Type._from_java(self._jvds.genotypeSignature())
+        return self._genotype_schema
+
+    @handle_py4j
+    def annotate_globals(self, **named_exprs):
+        exprs = []
+        named_exprs = {k: to_expr(v) for k, v in named_exprs.items()}
+        for k, v in named_exprs.items():
+            analyze(v, self, set(), set(), {'globals'}, False)
+            exprs.append('global.`{k}` = {v}'.format(k=k, v=v._ast.to_hql()))
+            self._check_field_name(k, set())
+        base, cleanup = self._process_joins(*named_exprs.values())
+        m = VariantDataset(self._hc, base._jvds.annotateGlobalExpr(",\n".join(exprs)))
+        return cleanup(m)
+
+    @handle_py4j
+    def annotate_rows(self, **named_exprs):
+        exprs = []
+        named_exprs = {k: to_expr(v) for k, v in named_exprs.items()}
+        for k, v in named_exprs.items():
+            analyze(v, self, {'row'}, {'column'}, set(self._all_fields), True)
+            exprs.append('va.`{k}` = {v}'.format(k=k, v=v._ast.to_hql().replace('%agg%', 'gs')))
```

Comment:
This will fail if (admittedly, unlikely) if the expression has an unexpected %agg% in it.  You should pass the aggregator as a argument to `to_hql`.

Reply:
I think I can do better, which is to write aggs as reference to __agg_uid__, then search the AST for references to __agg_uid__ and replace those.

Reply:
Then we don't allow __agg__ as a user variable?

Reply:
the __uid_1__ stuff could also cause collisions with user fields. I think we maybe should warn if people assign dunder columns.